main/src/ca/uwaterloo/flix/language/phase/unification/set/SetUnification.scala

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright 2024 Jonathan Lindegaard Starup
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase.unification.set
+
+import SetFormula.*
+
+object SetUnification {
+
+  /** A class to track progress during set unification. */
+  class Progress private(var somethingHappened: Boolean = false) {
+    /** Mark that progress has been made. */
+    def progress(): Unit = somethingHappened = true
+
+    /** Returns `true` if there has been progress. */
+    def hasProgressed: Boolean = somethingHappened
+  }
+
+  object Progress {
```

Comment:
companion object before class, by convention.

Reply:
Also: Maybe its better to be explicit where the progress object is created? I would do that and drop the apply method entirely.